lighthouse-core/audits/dobetterweb/uses-http2.js

```diff
@@ -43,67 +58,174 @@ class UsesHTTP2Audit extends Audit {
+    const simulationAfter = simulator.simulate(graph, {label: afterLabel, flexibleOrdering});
+
+    // Restore the original protocol after we've done our simulation
+    graph.traverse(node => {
+      if (node.type !== 'network') return;
+      const originalProtocol = originalProtocols.get(node.record.requestId);
+      if (originalProtocol === undefined) return;
+      node.record.protocol = originalProtocol;
+    });
+
+    const savingsOnOverallLoad = simulationBefore.timeInMs - simulationAfter.timeInMs;
+    const savingsOnTTI = Interactive.getLastLongTaskEndTime(simulationBefore.nodeTimings) -
+      Interactive.getLastLongTaskEndTime(simulationAfter.nodeTimings);
+    const savings = Math.max(savingsOnTTI, savingsOnOverallLoad);
+
+    // Round waste to nearest 10ms
+    return Math.round(Math.max(savings, 0) / 10) * 10;
+  }
+
+  /**
+   * Determines whether a network request is a "static resource" that would benefit from H2 multiplexing.
+   * XHRs, tracking pixels, etc generally don't benefit as much because they aren't requested en-masse
+   * for the same origin at the exact same time.
+   *
+   * @param {LH.Artifacts.NetworkRequest} networkRequest
+   * @return {boolean}
+   */
+  static isStaticAsset(networkRequest) {
+    if (!STATIC_RESOURCE_TYPES.has(networkRequest.resourceType)) return false;
+
+    // Resources from third-parties that are less than 100 bytes are usually tracking pixels, not actual resources.
+    // They can masquerade as static types though (gifs, documents, etc)
+    if (networkRequest.resourceSize < 100) {
+      const entity = ThirdPartyAudit.getEntitySafe(networkRequest.url);
+      if (entity) return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Determine the set of resources that aren't HTTP/2 but should be.
+   * We're a little conservative about what we surface for a few reasons:
+   *
+   *    - The simulator approximation of HTTP/2 is a little more generous than reality.
+   *    - There's a bit of debate surrounding HTTP/2 due to its worse performance in environments with high packet loss.
+   *    - It's something that you'd have absolutely zero control over with a third-party (can't defer to fix it for example).
+   *
+   * Therefore, we only surface requests that were...
+   *
+   *    - Served over HTTP/1.1 or earlier
+   *    - Served over an origin that serves at least 6 static asset requests
+   *      (if there aren't more requests than browser's max/host, multiplexing isn't as big a deal)
+   *
+   * @param {Array<LH.Artifacts.NetworkRequest>} networkRecords
+   * @return {Array<{url: string, protocol: string}>}
+   */
+  static determineNonHttp2Resources(networkRecords) {
+    /** @type {Array<{url: string, protocol: string}>} */
+    const nonHttp2Resources = [];
+
+    /** @type {Set<string>} */
+    const seenURLs = new Set();
+    /** @type {Map<string, Array<LH.Artifacts.NetworkRequest>>} */
+    const groupedByOrigin = new Map();
+    for (const record of networkRecords) {
```

Comment:
getting so sick of writing this pattern. would love a `groupBy(arr, cb)` function :P